### ISSUE TITLE
Add symbol font/language

### ISSUE THREAD

**Issue by LittleRichard** [LittleRichard/sormtger#4]
Much like #3, except symbols will need to be mapped to unicode characters so the raw output won't make much sense.  This will also need to define a lookup dict in python from the characters that are returned by the language mapping to a SetTypes enum.

Make the tesseract language files (and commit them), add the SetTypes enum, and add the lookup from tesseract output text to SetType.

**Comment by Panzer23** [comment]
Sets symbols that don't exist in paper:
Alpha (lea)
Beta (leb)
Unlimited (2ed)
Revised (3ed)
Summer Magic (psum)
Starter 2000 (s00)
masters edition (med)
masters edition II (me2)
Masters edition 3 (me3)
Masters edition IV(me4)
vintage amsters (vma)
tempest remastered (tpr)
legendary cube (xlcu)

**Comment by LittleRichard** [comment]
I've hit a wall with training tesseract on the symbols.  I can't get more than 85% of them to train, after a lot of tuning config params I don't quite understand, and the ones that did train perform... let's just say poorly.  As much as I'd like to apply the 'if it can do arabic it can do these symbols' theory, i can also see how they're significantly more complicated (and some of them are more or less identical).  

BUT, I have another idea:
https://en.wikipedia.org/wiki/Perceptual_hashing

**Comment by Panzer23** [comment]
Fair enough, structural similarity is probably the best best. The last link uses the same formula as the second.   MSE should be a good way too, if I can make the image sharp/similar enough.  which  I think I can do.

**Comment by Panzer23** [comment]
Also, what about this?
http://pastec.io/  not sure if you can feed it a library of images to search through, but it should be fast as fuck with only ~200 images to look through, especially if it has any kind of hashing.

**Comment by LittleRichard** [comment]
I don't think MSE will be a good fit because we'll end up with a fair amount of variance; it's a pixel-by-pixel comparison so if the symbol is shifted/skewed a bit instead of just color-adjusted like the example it will be returned as not-even close.  

The downside of these is that we'll need to come up with an entire suite of reference images.  One thing I need to figure out is if we can use a perfect symbol from my training TIFF or if we will have to gather them using the actual rig.

PastecIO: interesting, mostly because it's a newer algorithm 'ORB'.  Their library may be easier to use than OpenCV, but either way it's would seem to be a better method.  They emphasize that the true gain is that it's faster than SIFT, though I doubt given the image sizes we're talking about it'll make much difference.
http://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_feature2d/py_orb/py_orb.html

**Comment by Panzer23** [comment]
I'm actually not too worried about coming up with a library of images, it might be preferable anyway because of the changing backgrounds/colors/colors of the symbol, it would be a PITA, but Im used to having to do repetitive things to get a result.  Today for instance I am measuring water from 0C to 30C every five degrees....  WOOHOHOOOOO. But, I needed to do it so I am.  I may not have all the sets represented, but I'm pretty close and I can always buy more cards.

**Comment by yDgunz** [comment]
Generating the suite of reference images seems like an interesting challenge. I think this is what you're getting at above - can you use an image like http://magiccards.info/scans/en/vma/3.jpg or do you have to compare against an image generated by the rig? I would hope you can use the JPG, if you **have** to compare against images generated by the rig I'd be concerned about issues like different lighting conditions during sorting. 

I did a bit of Googling and it seems like you can't just download a full set of MTG card images, probably some licensing issues. That said, http://magiccards.info/ seems to have a full set (maybe? I don't know enough about MTG) so they must've gotten them from somewhere and they aren't endorsed/supported by Wizards.

**Comment by LittleRichard** [comment]
Generating the suite won't be too bad, it's accessing them efficiently that concerns me.  If the algorithm requires 2 candidates to make a match, it'll have to loop over every symbol (160 and counting) to find the best match...  

That's why the pHash thing is so interesting, it lets you boil down an image to an 8-byte number and then you can do basic 'get closest' type of operations to search the library.  They have a demo on their website that was good-not-great, but we have a small enough sample set that (as long as the rig's images are consistently placed) it might work.

It's possible that the reference images and close-crops from the card will be tiny enough to reasonably hold in memory, and small enough to be quickly sortable brute-force, but we must remember the target hardware is about half the speed of my desktop.

**Comment by yDgunz** [comment]
pHash looks promising. If that works out then you really only need to store the hashes, effectively solving your memory issue, right? And if you're only comparing the hashes I gotta imagine the processing power isn't an issue.

**Comment by LittleRichard** [comment]
More pHash-related stuff so i dont lose it:
http://www.hackerfactor.com/blog/?/archives/432-Looks-Like-It.html
http://www.hackerfactor.com/blog/index.php?/archives/529-Kind-of-Like-That.html

Perceptual hashing seems increasingly promising, though also very dependent on consistent symbol location

**Comment by Panzer23** [comment]
Well that just means my rig needs to be good I guess.  I think the way I have it laid out in my brain the only problem will be x OR y transforms, and as long as I can keep it to that, I think we should be ok.  Just adding a bit of height or width to the crops should not be that big of an issue.

**Comment by Panzer23** [comment]
@yDgunz you raise a good point about the lighting during sorting, I think I am gonna solve this with hardware.  There are a couple different finishes of card (foil/non-foil) and I am hoping to be able to distinguish them with a photosensor. I'm thinking the lighting needs to be consistent for that, so I was just gonna saturate the card to be sorted anyway, and block external light from the sorting spot

**Comment by Panzer23** [comment]
http://docs.opencv.org/2.4/doc/tutorials/imgproc/threshold/threshold.html#basic-threshold

do you think I should look into this sort of thing?

or is greyscale good enough?

**Comment by LittleRichard** [comment]
Thresholding is 100% the next move after grayscale, but once we crop down the image it'll be fast enough to do it in software so hold off for now.  It might screw up a perceptual hash anyway, especially the basic pHash one; dHash should be more okay with it.

I'm gonna take a crack at it a little this weekend, and hopefully have results of some kind Monday/Tuesday.

**Comment by Panzer23** [comment]
I'll upload some new samples that are a bit less sharp, but I have optimized the camera speed and capture time to that folder I gave you access too.

**Comment by LittleRichard** [comment]
Some interesting findings, jotting down my thoughts before I lose them.  I know this is a long post, with lots of stuff to read, but the results here are promising (with some caveats) and i'm really curious what you both think!  The output files are tuned to show minimal comparisons (since the X vs Y is bijective, it was redundant) but if it's easier to read I can upload the full comparison for each image file.

I grabbed a mix of images from magiccards.info representing a few different sets (some overlap, some not, you'll see the names in the output).  Here's roughly what the code does:
# load the images
# crop to a fixed box (this works just fine because the magiccards.info images are exactly the same size, so the box always yields exactly the symbol.  more on this in my notes)
# run the dHash. be impressed at the speed
# iterate through each image and compare it to each other by calculating the 'hamming distance', which is the number of bits the hash is off by.  0 is equivalent, greater than 30-or-so implies not a match, but there were some 25s that were matches (i think due to symbol-color-rarity differences, see below).
# print the output

1) this dhash thing is FAST, working with pretty tiny images but I think the performance will be comparable to the rig image sizes.  To boosts the matching likelihood, I may be able to 'smear' the pixel comparison window but it'll slow down the algorithm and make it less likely to identify complex symbols with minor differences (think m10-m15).  Either way, speed will not be an issue.

2) I cannot stress enough that no matter how clever the algorithm is, the chief limitation is that it's still a pixel-by-pixel comparison.  I tried a few crop-by-hands to get some small variances (even just a few pixels), and the performance degraded significantly.  I know i've said it several times, but i just proved it; unless we can get the crop box to be covering just about exactly the same spot on the card in every image it won't work.  I fear the 'just about' tolerance is going to be extremely small.

3) The rarity makes a difference because the symbol colors are different, and so a pixel-by-pixel of an all-black common symbol is very different than a silver uncommon symbol.  But that's actually an easy problem to solve, it won't be too bad to store the hash of each rarity of each set, all mapping back to the same set, and should greatly increase the match quality.

4) I didn't bother with aHash or pHash, they seem obviously worse to me.  Also, i had trouble installing the necessary libraries.  But we can still explore these if we want/need to.

5) I did try wHash, output also included here.  It is similarly performant, BUT given our lack of processing time limitations, it would be reasonable to run both and cross-reference the results.  I have a feeling it would boost the cases on the fence.

6) it is SO cool and ridiculously useful that the hashing boils the symbol down to a single number. I can already imagine the SQL required to do the search (sort by absolute_value(input_d_hash - table_of_hashes.d_hash))

[dhash_code.txt](https://github.com/LittleRichard/sormtger/files/490965/dhash_code.txt)
[dhash_output.txt](https://github.com/LittleRichard/sormtger/files/490964/dhash_output.txt)
[whash_output.txt](https://github.com/LittleRichard/sormtger/files/490970/whash_output.txt)

**Comment by LittleRichard** [comment]
Closing this issue.

I've merged a big change to master, including a fully-functional perceptual hash implementation.  We can discuss the image requirements for making it work, but the next challenge for symbol ID will be preparing the hash lookup data.

I'm pretty confident we can make this work, I have a few ideas for how we can get cute if the apparatus' tolerances aren't tight enough, but it's ready to demo tomorrow!
